be/src/storage/rowset/json_column_writer.cpp

```diff
@@ -0,0 +1,342 @@
+
+#include <sys/types.h>
+
+#include <algorithm>
+#include <cstdint>
+#include <memory>
+#include <sstream>
+#include <string>
+#include <utility>
+
+#include "column/column.h"
+#include "column/column_helper.h"
+#include "column/column_viewer.h"
+#include "column/json_column.h"
+#include "column/nullable_column.h"
+#include "column/vectorized_fwd.h"
+#include "common/config.h"
+#include "common/status.h"
+#include "gen_cpp/segment.pb.h"
+#include "gutil/casts.h"
+#include "storage/rowset/column_writer.h"
+#include "types/logical_type.h"
+#include "util/json_util.h"
+#include "velocypack/vpack.h"
+
+namespace starrocks {
+
+class FlatJsonColumnWriter final : public ColumnWriter {
+public:
+    FlatJsonColumnWriter(const ColumnWriterOptions& opts, const TypeInfoPtr& type_info, WritableFile* wfile,
+                         std::unique_ptr<ScalarColumnWriter> json_writer);
+
+    ~FlatJsonColumnWriter() override = default;
+
+    Status init() override { return _json_column_writer->init(); };
+
+    Status append(const Column& column) override;
+    Status append(const ColumnPtr& column) override;
+
+    Status finish_current_page() override;
+
+    uint64_t estimate_buffer_size() override;
+
+    Status finish() override;
+
+    Status write_data() override;
+    Status write_ordinal_index() override;
+    Status write_zone_map() override;
+    Status write_bitmap_index() override;
+    Status write_bloom_filter_index() override;
+    ordinal_t get_next_rowid() const override { return _json_column_writer->get_next_rowid(); }
+
+    bool is_global_dict_valid() override { return _json_column_writer->is_global_dict_valid(); }
+
+    uint64_t total_mem_footprint() const override { return _json_column_writer->total_mem_footprint(); }
+
+private:
+    void _flat_column(std::vector<ColumnPtr>& json_datas);
+
+private:
+    std::unique_ptr<ScalarColumnWriter> _json_column_writer;
+    ColumnMetaPB* _json_meta;
+    WritableFile* _wfile;
+
+    std::vector<ColumnPtr> _json_datas;
+
+    std::vector<std::unique_ptr<ColumnWriter>> _flat_writers;
+    std::vector<std::string> _flat_paths;
+    std::vector<ColumnPtr> _flat_columns;
+};
+
+FlatJsonColumnWriter::FlatJsonColumnWriter(const ColumnWriterOptions& opts, const TypeInfoPtr& type_info,
+                                           WritableFile* wfile, std::unique_ptr<ScalarColumnWriter> json_writer)
+        : ColumnWriter(std::move(type_info), opts.meta->length(), opts.meta->is_nullable()),
+          _json_column_writer(std::move(json_writer)),
+          _json_meta(opts.meta),
+          _wfile(wfile) {}
+
+Status FlatJsonColumnWriter::append(const Column& column) {
+    return Status::InternalError("json column writer not support append reference interface");
+}
+
+Status FlatJsonColumnWriter::append(const ColumnPtr& column) {
+    _json_datas.emplace_back(column);
+    return Status::OK();
+}
+
+void FlatJsonColumnWriter::_flat_column(std::vector<ColumnPtr>& json_datas) {
```

Comment:
the flatten logic could be separated with ColumnWriter

Reply:
the flatten logic has been move to JSONFlater. here is some strategies to determine whether flatten is necessary, and it's only used in write process